docs/manual/working/scalaGuide/main/tests/code/ScalaFunctionalTestSpec.scala

```diff
@@ -39,20 +36,19 @@ class ScalaFunctionalTestSpec extends MixedPlaySpec with Results {
       .build()
     // #scalafunctionaltest-application
 
-    val applicationWithRouter = new GuiceApplicationBuilder().router(Router.from {
-      case GET(p"/Bob") =>
-        Action {
-          Ok("Hello Bob") as "text/html; charset=utf-8"
+    val applicationWithRouter = GuiceApplicationBuilder()
+      .appRoutes(app => {
+        case ("GET", "/Bob") => app.injector.instanceOf(classOf[DefaultActionBuilder]) {
```

Comment:
Not the best idiom. Maybe we need to overload `appRoutes` to have something like:

```scala
def appRoutes(routes: ControllerComponents => PartialFunction[(String, String), Handler]): GuiceApplicationBuilder
```

But for now, this is how we avoid global state while using `GuiceApplicationBuilder` and configuring a `Router`.

Reply:
That's fine for a milestone build, but I agree we'll want something improved.

FYI, in the Play integration I realised there are a lot of different ways that we might want to make an application. I factored out all the recipes into an "ApplicationFactory". Then you can just provide an ApplicationFactory and all the different nice methods are available:

https://github.com/playframework/playframework/blob/084a7ab635934a4ef7ac07e212d044a36a516d85/framework/src/play-integration-test/src/test/scala/play/it/test/ApplicationFactory.scala#L23-L48